File: examples/canonical-aws-lambda/apigw-stripe-webhook/typescript/tests/replay/test_replay.js
const assert = require("assert");
const crypto = require("crypto");
const { handler } = require("../../src/handler.js");

async function main() {
  const workflows = [];
  const stored = new Set();
  const body = JSON.stringify({ id: "evt_replay", type: "invoice.paid" });
  const secret = "local-signing-secret";
  const signature = crypto.createHmac("sha256", secret).update(body).digest("hex");
  const clients = {
    getSigningSecret: async () => secret,
    storeRawEvent: async (id) => {
      if (stored.has(id)) return false;
      stored.add(id);
      return true;
    },
    startWorkflow: async (id) => workflows.push(id)
  };
  const event = { body, headers: { "stripe-signature": `v1=${signature}` }, requestContext: { requestId: "req_replay" } };
  const first = await handler(event, clients);
  const second = await handler(event, clients);
  assert.equal(first.statusCode, 202);
  assert.equal(second.statusCode, 202);
  assert.equal(stored.size, 1);
  assert.equal(workflows.length, 1);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: examples/canonical-aws-lambda/apigw-stripe-webhook/python/tests/replay/test_replay.py
import pathlib
import sys
import unittest

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parents[2] / "src"))
sys.path.insert(0, str(pathlib.Path(__file__).resolve().parents[1] / "smoke"))

from handler import lambda_handler
from test_handler import FakeClients, event


class ReplayTest(unittest.TestCase):
    def test_same_event_starts_one_workflow(self):
        clients = FakeClients()
        first = lambda_handler(event("evt_replay"), clients=clients)
        second = lambda_handler(event("evt_replay"), clients=clients)
        self.assertEqual(first["statusCode"], 202)
        self.assertEqual(second["statusCode"], 202)
        self.assertEqual(len(clients.stored), 1)
        self.assertEqual(clients.workflows, ["evt_replay"])


if __name__ == "__main__":
    unittest.main()
